### PR TITLE
fix: normalize waveforms before torchaudio.save to avoid saturated WAVs on torchaudio 2.9+

### LINKED ISSUE #724: Audio save path is incompatible with TorchAudio 2.9+ and can silently saturate generated WAV files
## Summary

The current IndexTTS audio save path converts the generated normalized waveform to PCM-scale values and then passes an `int16` tensor to `torchaudio.save()`.

This works with the TorchAudio 2.8 save implementation used by the currently supported IndexTTS environment.

However, beginning with TorchAudio 2.9, `torchaudio.save()` delegates encoding to TorchCodec. In that path, PCM-scale `int16` input is converted to `float32` without normalization and is then interpreted as audio that should be in `[-1, 1]`.

The save operation succeeds without an exception or warning, but the resulting WAV is silently saturated. Almost every non-zero sample becomes either negative or positive full scale, destroying the amplitude information.

This report is not claiming that the current official IndexTTS installation is broken. The official dependencies currently pin Torch and TorchAudio to `2.8.*`, which avoids the new save path. The problem appears when this code is used with TorchAudio 2.9 or later, including upgraded, modified, or downstream-packaged environments.

## Current IndexTTS behavior

IndexTTS converts the generated waveform to PCM scale before saving:

```python
wav = torch.clamp(32767 * wav, -32767.0, 32767.0)
```

It later saves that PCM-scale waveform as an integer tensor:

```python
torchaudio.save(
    output_path,
    wav.type(torch.int16),
    sampling_rate,
)
```

This pattern is present in both:

* `indextts/infer.py`
* `indextts/infer_v2.py`

Current source:

https://github.com/index-tts/index-tts/blob/main/indextts/infer.py

https://github.com/index-tts/index-tts/blob/main/indextts/infer_v2.py

The current project dependencies pin:

```toml
torch==2.8.*
torchaudio==2.8.*
```

https://github.com/index-tts/index-tts/blob/main/pyproject.toml

Therefore, the official `uv` environment currently avoids the failure. The save code itself nevertheless becomes incompatible if TorchAudio is updated to 2.9 or later.

## Why it fails

With TorchAudio 2.9 and later, the compatibility wrapper effectively performs:

```python
if src.dtype != torch.float32:
    src = src.float()
```

For example, a PCM int16 value of:

```text
26212
```

becomes:

```text
26212.0
```

rather than a normalized value near:

```text
0.8
```

TorchCodec expects float audio samples in `[-1, 1]`. The out-of-range waveform is accepted and clipped during encoding.

## Minimal reproduction of the IndexTTS save pattern

No IndexTTS model or checkpoint is required to reproduce the save-boundary problem.

```python
from array import array
import sys
import wave

import torch
import torchaudio


sample_rate = 24000
num_samples = sample_rate

t = torch.arange(num_samples, dtype=torch.float32) / sample_rate

# Synthetic normalized model-like output.
normalized = (
    0.72 * torch.sin(2 * torch.pi * 440.0 * t)
    + 0.08 * torch.sin(2 * torch.pi * 997.0 * t)
).clamp(-0.8, 0.8).unsqueeze(0).contiguous()

# Equivalent to the current IndexTTS scaling step.
pcm_scale_wav = torch.clamp(
    32767.0 * normalized,
    -32767.0,
    32767.0,
)

# Current IndexTTS-style save.
torchaudio.save(
    "indextts_current_save.wav",
    pcm_scale_wav.to(torch.int16),
    sample_rate,
)

# Compatible save using normalized float32.
torchaudio.save(
    "indextts_normalized_save.wav",
    (pcm_scale_wav / 32767.0).to(torch.float32),
    sample_rate,
)


def inspect_pcm16(path):
    with wave.open(path, "rb") as wav_file:
        raw = wav_file.readframes(wav_file.getnframes())

    samples = array("h")
    samples.frombytes(raw)

    if sys.byteorder != "little":
        samples.byteswap()

    values = list(samples)

    fullscale_fraction = sum(
        value <= -32768 or value >= 32767
        for value in values
    ) / len(values)

    return {
        "min": min(values),
        "max": max(values),
        "unique_values": len(set(values)),
        "fullscale_fraction": fullscale_fraction,
    }


print(
    "current save:",
    inspect_pcm16("indextts_current_save.wav"),
)

print(
    "normalized save:",
    inspect_pcm16("indextts_normalized_save.wav"),
)
```

## Observed result

Environment:

* Python 3.12.13
* PyTorch 2.11.0+cpu
* TorchAudio 2.11.0+cpu
* TorchCodec 0.15.0+cpu

Current IndexTTS-style save:

```text
{
    'min': -32768,
    'max': 32767,
    'unique_values': 3,
    'fullscale_fraction': 0.999875
}
```

The saved PCM contains effectively only:

```text
-32768, 0, 32767
```

The sign of the waveform is retained, but nearly all amplitude information is lost.

Normalized float32 save:

```text
{
    'min': -26213,
    'max': 26213,
    'unique_values': 18849,
    'fullscale_fraction': 0.0
}
```

## Full reproduction notebook

A public, executable CPU-only notebook is available here:

https://huggingface.co/datasets/John6666/forum3/blob/main/torchaudio_torchcodec_audio_save_reproducer/torchaudio_torchcodec_audio_save_reproducer_final.ipynb

It compares:

* `torchaudio.save()`
* `torchaudio.save_with_torchcodec()`
* direct TorchCodec `AudioEncoder`
* normalized float32 input
* PCM-scale int16 input
* PCM-scale float32 input
* out-of-range float32 input

All automated observations in the notebook evaluate to `OBSERVED`.

## Suggested fix direction

A minimal compatibility fix would be to normalize the existing PCM-scale waveform immediately before file saving:

```python
torchaudio.save(
    output_path,
    (wav.to(torch.float32) / 32767.0).clamp(-1.0, 1.0),
    sampling_rate,
)
```

Because IndexTTS already clamps `wav` to `[-32767, 32767]`, division by `32767.0` restores the expected normalized range.

This approach has several advantages:

* It does not require checking the TorchAudio version.
* It works with the current TorchAudio 2.8 environment.
* It works with the TorchCodec-backed save path in TorchAudio 2.9 and later.
* It does not require changing the existing PCM-scale streaming output.
* It does not require changing the existing int16 NumPy return value used by the UI.

The exact WAV encoding or subtype should remain a maintainer decision. The important requirement is that normalized float samples, rather than PCM-scale integer samples, reach the TorchCodec-backed save path.

A larger future refactor could preserve the normalized model waveform until the final output boundary and create PCM-scale copies only for APIs that specifically require them. That is not necessary for the minimal compatibility fix.

## Upstream issues

TorchAudio compatibility wrapper:

https://github.com/pytorch/audio/issues/4211

TorchCodec value-range validation:

https://github.com/meta-pytorch/torchcodec/issues/1576

The TorchAudio issue covers the legacy integer-input compatibility regression.

The TorchCodec issue separately covers the fact that `AudioEncoder` validates the required dtype but does not reject or warn about float32 values outside the documented `[-1, 1]` range.

## Possibly related IndexTTS reports

The following reports describe loud or heavily corrupted output and may have a similar audible failure mode:

* #540 — Electric and loud static in audio
  https://github.com/index-tts/index-tts/issues/540

* #496 — Noisy output after moving IndexTTS 1.5 and 2.0 to NVIDIA DGX Spark
  https://github.com/index-tts/index-tts/issues/496

Issue #540 reports that every generation sounds similar and very loud.

Issue #496 reports clean output on a Mac mini but unusably noisy output after moving both IndexTTS 1.5 and 2.0 to a different environment.

These reports do not include enough dependency information or pre-save/post-save waveform statistics to establish that they have the same root cause. They are linked only as possibly related audible symptoms.

## Workarounds

For users running TorchAudio 2.9 or later:

```python
torchaudio.save(
    output_path,
    (wav.to(torch.float32) / 32767.0).clamp(-1.0, 1.0),
    sampling_rate,
)
```

Alternatively, using the officially pinned TorchAudio 2.8 environment avoids this specific compatibility failure.

> **AI assistance disclosure:** AI assistance was used to organize and edit this report and the accompanying reproduction notebook. I ran the experiments and reviewed all results and technical claims. The reproduction, observations, and conclusions above reflect my own review.

### PR DESCRIPTION
**Pull request by Souptik96**
## Description

Fixes #724.

`torchaudio.save()` was being handed a **PCM-scale `int16`** tensor at every save site. From torchaudio 2.9 onward, `save()` delegates to TorchCodec, and its compatibility shim does exactly this ([2.9 source](https://github.com/pytorch/audio/blob/v2.9.0/src/torchaudio/_torchcodec.py#L292-L293)):

```python
if src.dtype != torch.float32:
    src = src.float()
```

No rescaling. A sample of `26212` becomes `26212.0`, and `AudioEncoder` then interprets it as audio that should be in `[-1, 1]`. Everything is clipped to a rail. No exception, no warning — just a saturated file.

This PR normalizes to float32 `[-1, 1]` before saving, which is the one input range every torchaudio version interprets the same way.

### Why the encoding is pinned too

The workaround in the issue (normalize only) fixes the amplitude but introduces a second, quieter regression on the currently pinned torchaudio 2.8: below 2.9 the WAV subtype is derived from the **input dtype** ([`_get_subtype_for_wav`](https://github.com/pytorch/audio/blob/v2.8.0/src/torchaudio/_backend/soundfile_backend.py#L239-L252)), so float32 input silently produces a **32-bit IEEE float WAV** instead of the 16-bit PCM this project has always written. I measured that, and the regression test covers it:

```
torchaudio 2.8.0, normalize only, no encoding pinned
  -> IEEE_float32   (expected PCM_int16)
```

So `save_pcm_wav()` passes `encoding="PCM_S", bits_per_sample=16` on versions that honour them, and omits them on 2.9+ where they are [documented as ignored and warn if supplied](https://github.com/pytorch/audio/blob/v2.9.0/src/torchaudio/__init__.py#L117-L120) (TorchCodec's WAV default is already 16-bit signed PCM). Net effect: byte-identical container format on every version tested, and no new warnings.

### Changes

- `indextts/utils/common.py` — new `save_pcm_wav()` + `PCM16_MAX`, next to the existing `load_audio()`.
- Routed all six save sites through it:
  - `indextts/infer.py` (`infer_fast`, `infer`)
  - `indextts/infer_v2.py`
  - `indextts/infer_v2_5.py` (batch-concat path and main path)
  - `backends/trt/infer.py` (3 sites), `backends/trt/pipeline/pipeline.py` (`_save_wav`)

The issue only named `infer.py` and `infer_v2.py`; `infer_v2_5.py` and `backends/trt/` have the identical bug from the identical cause, so they are fixed in the same pass rather than left behind. `backends/trt/` pins `torchaudio>=2.7.1,<2.8`, which is also verified below.

The int16 NumPy return value used by the WebUI and the PCM-scale streaming output are **unchanged** — only the file-writing boundary moved.

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that changes existing behaviour)
- [ ] Documentation / comments only
- [ ] Build / CI / dependency change
- [ ] Refactor (no behaviour change)

## How has this been tested?

New file `tests/test_audio_save.py` — 10 CPU-only tests, no checkpoints, so they run in the existing `not gpu` CI job. The WAV reader in the test parses RIFF chunks by hand rather than using an audio library, so the same normalization bug cannot cancel itself out on the read side.

```bash
uv run --extra test pytest -m "not gpu"
```

**Full `not gpu` suite (torchaudio 2.8.0, the pinned version):**

```
163 passed, 22 deselected, 30 subtests passed in 25.17s
```

**The new tests across four torchaudio versions** (real CPU installs, TorchCodec + FFmpeg 7.1 present for 2.9/2.11):

| python | torchaudio | torchcodec | result |
|---|---|---|---|
| 3.10.20 | 2.8.0 | — | 10 passed |
| 3.12.13 | 2.7.1 | — | 10 passed |
| 3.12.13 | 2.8.0 | — | 10 passed |
| 3.12.13 | 2.9.0 | 0.9.1 | 10 passed |
| 3.12.13 | 2.11.0 | 0.16.0 | 10 passed |

The 3.10 row matches what `ci.yml` provisions. 2.7.1 covers the `backends/trt/` pin. The tests need a torchaudio save backend, which `soundfile` provides; it is a locked transitive dependency of both `librosa` and `descript-audiotools`, so it is always present.

**Negative control** — reverting `save_pcm_wav()` to the old `torchaudio.save(path, wav.type(torch.int16), sr)`:

| torchaudio | before fix | after fix |
|---|---|---|
| 2.8.0 | **5 failed**, 5 passed | 10 passed |
| 2.9.0 | **7 failed**, 3 passed | 10 passed |

Note the suite fails on 2.8 as well, even though 2.8's *output file* is fine. That is deliberate: the contract tests assert what is handed to `torchaudio.save()` rather than only inspecting the result, so CI (which runs 2.8) actually guards this instead of going green until someone bumps the pin.

Reverting to the issue's normalize-only workaround instead fails 3 tests on 2.8, catching the `IEEE_float` subtype change.

## Screenshots / output (if relevant)

Input is a 1 s 440 Hz + 997 Hz tone at 0.8 full scale, run through `torch.clamp(32767 * wav, -32767., 32767.)` exactly as the inference code does, then saved and re-read. Expected: `PCM_int16`, peak ≈ 26212, many distinct levels, 0 clipped frames.

```
torchaudio 2.7.1+cpu
  BEFORE (old code)      PCM_int16    peak=26211  levels=18846  clipped=0.000000  err=5.5e-05
  AFTER  (save_pcm_wav)  PCM_int16    peak=26212  levels=18898  clipped=0.000000  err=3.1e-05

torchaudio 2.8.0+cpu
  BEFORE (old code)      PCM_int16    peak=26211  levels=18846  clipped=0.000000  err=5.5e-05
  AFTER  (save_pcm_wav)  PCM_int16    peak=26212  levels=18898  clipped=0.000000  err=3.1e-05

torchaudio 2.9.0+cpu
  BEFORE (old code)      PCM_int16    peak=32767  levels=3      clipped=0.999875  err=0.999874   <-- broken
  AFTER  (save_pcm_wav)  PCM_int16    peak=26213  levels=18849  clipped=0.000000  err=1.5e-05

torchaudio 2.11.0+cpu
  BEFORE (old code)      PCM_int16    peak=32767  levels=3      clipped=0.999875  err=0.999874   <-- broken
  AFTER  (save_pcm_wav)  PCM_int16    peak=26213  levels=18849  clipped=0.000000  err=1.5e-05
```

`levels=3` on the broken rows means the entire waveform collapsed onto `{-32768, 0, 32767}` — this reproduces @John6666cat's reported figures (`unique_values: 3`, `fullscale_fraction: 0.999875`) exactly.

## Not verified

`backends/trt/` cannot be executed here — it needs TensorRT-LLM engines and an NVIDIA GPU. Those four call sites are changed by code review only. They are mechanical substitutions of the same pattern, and the tensors reaching them are documented and asserted as PCM-scale int16 in that code (`streaming.py::_to_int16` is `audio * 32767`, `pipeline.py` line 646 comment "int16-scale", and the `AudioOutput.audio: np.ndarray # int16` annotation), but a maintainer with TRT hardware should confirm one generated file.

### CODE REVIEW

**Comment by John6666cat** [comment]
Thanks for picking this up. This matches what I found in #724.
Nice catch on the TorchAudio 2.8 PCM16 behavior too. **I had missed that**.😂
Thanks for taking this further.
